File: apps/agents/src/open-canvas/nodes/generate-path/dynamic-determine-path.ts
import {
  ROUTE_QUERY_PROMPT,
  ROUTE_QUERY_OPTIONS_HAS_ARTIFACTS,
  ROUTE_QUERY_OPTIONS_NO_ARTIFACTS,
  CURRENT_ARTIFACT_PROMPT,
  NO_ARTIFACT_PROMPT,
} from "../../prompts.js";
import { OpenCanvasGraphAnnotation } from "../../state.js";
import { createContextDocumentMessagesOpenAI as createContextDocumentMessages } from "../../../lib/context-docs";
import { getModelFromConfigLocal } from "../../../lib/model-config.local";
import { LangGraphRunnableConfig } from "@langchain/langgraph";
import { getArtifactContent } from "@opencanvas/shared/utils/artifacts";
import z from "zod";
import { BaseMessage } from "@langchain/core/messages";

interface DynamicDeterminePathParams {
  state: typeof OpenCanvasGraphAnnotation.State;
  newMessages: BaseMessage[];
  config: LangGraphRunnableConfig;
}

/**
 * Dynamically determines the path to take using an LLM.
 */
async function dynamicDeterminePathFunc({
  state,
  newMessages,
  config,
}: DynamicDeterminePathParams) {
  const currentArtifactContent = state.artifact
    ? getArtifactContent(state.artifact)
    : undefined;

  // Call model and decide if we need to respond to a users query, or generate a new artifact
  const formattedPrompt = ROUTE_QUERY_PROMPT.replace(
    "{artifactOptions}",
    currentArtifactContent
      ? ROUTE_QUERY_OPTIONS_HAS_ARTIFACTS
      : ROUTE_QUERY_OPTIONS_NO_ARTIFACTS
  )
    .replace(
      "{recentMessages}",
      state._messages
        .slice(-3)
        .map((message: BaseMessage) => `${message.getType()}: ${message.content}`)
        .join("\n\n")
    )
    .replace(
      "{currentArtifactPrompt}",
      currentArtifactContent
        ? CURRENT_ARTIFACT_PROMPT
        : NO_ARTIFACT_PROMPT
    );

  const artifactRoute = currentArtifactContent
    ? "rewriteArtifact"
    : "generateArtifact";

  const model = await getModelFromConfigLocal();

  const schema = z.object({
    route: z
      .enum(["replyToGeneralInput", artifactRoute])
      .describe("The route to take based on the user's query."),
  });

  const modelWithTool = model.bindTools(
    [
      {
        name: "route_query",
        description: "The route to take based on the user's query.",
        schema,
      },
    ],
    {
      tool_choice: "route_query",
    }
  );

  const contextDocumentMessages = await createContextDocumentMessages(state._messages);
  const result = await modelWithTool.invoke([
    ...contextDocumentMessages,
    ...(newMessages.length ? newMessages : []),
    {
      role: "user",
      content: formattedPrompt,
    },
  ]);

  if (!result.tool_calls || result.tool_calls.length === 0) {
    throw new Error("No tool calls found in the result");
  }
  const args = result.tool_calls[0].args;
  if (!args) {
    throw new Error("No args found in the tool call");
  }
  return args as z.infer<typeof schema>;
}

export const dynamicDeterminePath = dynamicDeterminePathFunc;

File: agents-py/agents/src/open_canvas/nodes/generate_path/dynamic_determine_path.py
from typing import Dict, Any, List
from pydantic import BaseModel, Field
from langchain_core.messages import BaseMessage, HumanMessage
from langchain_core.runnables import RunnableConfig
from agents.src.open_canvas.prompts import (
    ROUTE_QUERY_PROMPT,
    ROUTE_QUERY_OPTIONS_HAS_ARTIFACTS,
    ROUTE_QUERY_OPTIONS_NO_ARTIFACTS,
    CURRENT_ARTIFACT_PROMPT,
    NO_ARTIFACT_PROMPT
)
from agents.src.utils import (
    format_artifact_content_with_template,
    get_model_from_config,
    create_context_document_messages,
    get_context_documents
)
from shared.src.utils.artifacts import get_artifact_content
from langsmith import traceable

class RouteSchema(BaseModel):
    """Schema for routing response."""
    route: str = Field(
        ..., 
        description="The route to take based on the user's query.",
        enum=["reply_to_general_input", "rewrite_artifact", "generate_artifact"]
    )

@traceable(name="dynamic_determine_path")
async def dynamic_determine_path(
    state: Dict[str, Any],
    new_messages: List[BaseMessage],
    config: RunnableConfig
) -> Dict[str, str]:
    """Determine the next route based on user input and current state."""
    try:
        # Initialize model
        model = await get_model_from_config(config, {"temperature": 0, "is_tool_calling": True})
        
        # Get current artifact content
        current_artifact = state.get("artifact")
        current_artifact_content = get_artifact_content(current_artifact) if current_artifact else None
        
        # Format prompt components
        artifact_options = (
            ROUTE_QUERY_OPTIONS_HAS_ARTIFACTS if current_artifact_content 
            else ROUTE_QUERY_OPTIONS_NO_ARTIFACTS
        )
        
        recent_messages = "\n\n".join(
            f"{msg.type}: {msg.content}" 
            for msg in state.get("_messages", [])[-3:]
        )
        
        current_artifact_prompt = (
            format_artifact_content_with_template(
                CURRENT_ARTIFACT_PROMPT,
                current_artifact_content
            ) if current_artifact_content else NO_ARTIFACT_PROMPT
        )
        
        # Format final prompt
        formatted_prompt = ROUTE_QUERY_PROMPT.format(
            artifact_options=artifact_options,
            recent_messages=recent_messages,
            current_artifact_prompt=current_artifact_prompt
        )

        # Get context documents
        context_docs = await create_context_document_messages(
            documents=await get_context_documents(config),
            provider="openai",  # Simplified for now
            model_name="gpt-4"
        )
        
        # Prepare messages
        messages = [
            *[HumanMessage(content=msg) for msg in context_docs],
            *(new_messages or []),
            HumanMessage(content=formatted_prompt)
        ]

        # Get route from model
        model_with_tools = model.bind_tools(
            tools=[{
                "name": "route_query",
                "description": "Determine the next route based on user input",
                "schema": RouteSchema.model_json_schema()
            }],
            tool_choice="route_query"
        )
        
        response = model_with_tools.invoke(messages)
        
        # Extract route from response
        if response.tool_calls:
            tool_call = response.tool_calls[0]
            if hasattr(tool_call, "function"):
                import json
                args = json.loads(tool_call.function.arguments)
                if args.get("route"):
                    return {"next": args["route"]}

        # Default fallback
        return {"next": "reply_to_general_input"}
        
    except Exception as e:
        print(f"Error in dynamic_determine_path: {e}")
        return {"next": "reply_to_general_input"} 